### PR TITLE
🕷️ 854 Bug/ Lost to Follow Up QA

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

- Refactors + checks different fields for different entities
- Updated handling for Primary Diagnosis
- Addresses issues mentioned in:
https://github.com/icgc-argo/argo-clinical/issues/854#issuecomment-1724284915

**Type of Change**

- [x] Bug
- [x] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/submission/validation-clinical/donor.ts [hunk 1]
```diff
@@ -138,81 +145,73 @@ const crossFileValidator = async (
...
-        const totalTreatmentTime = treatmentStartInterval + treatmentDuration;
-
-        return totalTreatmentTime > lostToFollowUpInterval;
+      const lostToFollowUpClinicalInfo = donorClinicalEventIdMatch.clinicalInfo;
+
+      const lostToFollowUpStartInterval =
+        Number(
+          lostToFollowUpClinicalInfo.interval_of_followup ||
+            lostToFollowUpClinicalInfo.treatment_start_interval ||
+            lostToFollowUpClinicalInfo.specimen_acquisition_interval,
```
Comment (demariadaniel) [inline src/submission/validation-clinical/donor.ts:154]: Need to check across different keys for different entities; was previously only checking `interval_of_followup`

File: src/submission/submission-error-messages.ts [hunk 2]
```diff
@@ -124,13 +124,13 @@ const ERROR_MESSAGES: { [key: string]: (errorData: any) => string } = {
   INVALID_LOST_TO_FOLLOW_UP_ID: ({ info: { lost_to_followup_after_clinical_event_id } }) =>
     `The identifier '${lost_to_followup_after_clinical_event_id}' submitted in the 'lost_to_followup_after_clinical_event_id' field does not exist in your clinical submission.`,
   INVALID_SUBMISSION_AFTER_LOST_TO_FOLLOW_UP: ({
-    info: {
-      lost_to_followup_after_clinical_event_id,
-      interval_of_followup,
-      submitter_treatment_id,
-    },
+    info: { lost_to_followup_after_clinical_event_id, interval_of_followup, treatment_id },
```
Comment (demariadaniel) [inline src/submission/submission-error-messages.ts:127]: treatment_id is any of `submitter_treatment_id || submitter_specimen_id || submitter_follow_up_id`

File: src/submission/validation-clinical/donor.ts [hunk 3]
```diff
@@ -138,81 +145,126 @@ const crossFileValidator = async (
...
+          ),
+        );
+      }
 
-        const lostToFollowUpPrimaryDiagnosisMatch = primaryDiagnoses?.find(
-          primaryDiagnosisRecord =>
-            primaryDiagnosisRecord?.clinicalInfo?.submitter_primary_diagnosis_id ===
-            lostToFollowUpDiagnosisId,
-        )?.clinicalInfo;
+      // Filter Primary Diagnosis records with Age at Diagnosis greater than age at Lost To Follow Up
```
Comment (demariadaniel) [inline src/submission/validation-clinical/donor.ts:224]: Primary Diagnosis contains `age` but not `interval` data (years vs days) and is checked separately

File: src/submission/validation-clinical/donor.ts [hunk 4]
```diff
@@ -138,81 +145,126 @@ const crossFileValidator = async (
...
+
+          const specimenInterval =
+            typeof specimen_acquisition_interval === 'number'
+              ? specimen_acquisition_interval + (Number(specimen_duration) || 0)
+              : 0;
+
+          return (
+            treatmentInterval > lostToFollowUpInterval || specimenInterval > lostToFollowUpInterval
+          );
+        })
```
Comment (joneubank) [inline src/submission/validation-clinical/donor.ts:185]: I was surprised that TS was fine letting you deconstruct these properties from objects of two different types... but of course we don't know the structure of ClinicalInfo objects, they are basically just JSON objects, so this works...

In any case, I see what you are doing here but its a bit confusing handling entities of two different types the same way. Would you consider writing some helper methods to clean this a bit? 

`getTreatmentInterval(treatment: DeepReadonly<Treatment>): number` and `getSpecimenInterval(DeepReadonly<Specimen>): number`

and from that:

```ts
const invalidClinicalIntervalRecords = [
    ...treatments.filter(treatment => getTreatmentInterval(treatment) > lostToFollowUpInterval),
    ...specimens.filter(specimen => getSpecimenInterval(treatment) > lostToFollowUpInterval)
].map(record => record.clinicalInfo);
```

**Review by joneubank — APPROVED** [review]
Looks like you solved the problem. My comment is a suggestion to clarify what a tricky filter is doing.

File: src/submission/validation-clinical/donor.ts (hunk 4 above)
Comment (demariadaniel) [reply src/submission/validation-clinical/donor.ts:185]: Yes, agreed. Now that the logic is all sorted I can straighten this out.

Comment (demariadaniel) [reply src/submission/validation-clinical/donor.ts]: Pushed an update

File: src/submission/validation-clinical/donor.ts [hunk 5]
```diff
@@ -96,6 +96,24 @@ function checkTimeConflictWithSpecimens(
...
+};
+
+const getSpecimenInterval = (clinicalInfo: DeepReadonly<ClinicalInfo>) => {
+  const { specimen_acquisition_interval, specimen_duration } = clinicalInfo;
+  const specimenInterval =
+    typeof specimen_acquisition_interval === 'number'
+      ? specimen_acquisition_interval + (Number(specimen_duration) || 0)
+      : 0;
+  return specimenInterval;
+};
```
Comment (joneubank) [inline src/submission/validation-clinical/donor.ts:115]: `ClinicalInfo` has no shape information, update these to accept `Treamtent` or `Specimen` as the generic type. You can pull their clinicalInfo out within the method.

File: src/submission/validation-clinical/donor.ts [hunk 6]
```diff
@@ -163,27 +181,14 @@ const crossFileValidator = async (
...
-        })
-        .map(record => record.clinicalInfo);
+      const invalidClinicalIntervalRecords = [
+        ...treatments
+          .map(treatmentRecord => treatmentRecord.clinicalInfo)
+          .filter(clinicalInfo => getTreatmentInterval(clinicalInfo) > lostToFollowUpInterval),
+        ...specimens
+          .map(specimenRecord => specimenRecord.clinicalInfo)
+          .filter(clinicalInfo => getSpecimenInterval(clinicalInfo) > lostToFollowUpInterval),
+      ];
```
Comment (joneubank) [inline src/submission/validation-clinical/donor.ts:191]: Its not necessary to run `.map` here, just make the `getTreatmentInterval` method accept the treatment.

**Review by joneubank — APPROVED** [review]

**Pull request merged**